### ISSUE TITLE
[Finding] Deterministic harness bridge tokens still require caller-side resume-state rehydration

### ISSUE THREAD

**Issue by teamleaderleo** [teamleaderleo/fieldwork#805]
State: `observed`

Programme: #13  
Target hub: #2  
Source scout: #783  
Target characterization: `teamleaderleo/ai#51`  
Execution carrier: `teamleaderleo/ai#52`  
Exact target base: `fc3baaf1ff547efdfcc5cb5a5ee35ed72b8a284c`  
Current public head checked: `cfc587bdfd8fd1996dd902edd14143be6e034baf`  
Claim scope: interface  
Upstream contact authorized: `false`

## In simple words

The new deterministic `mintBridgeToken(sandboxId)` setting gives an application enough information to derive the same bridge bearer token again. Live resume state still contains that secret as a required `bridge.token`, though, and `HarnessAgent.createSession()` validates the state before the adapter gets a chance to derive anything.

So an application that wants durable resume storage without retaining the bridge bearer value must implement its own two-step serializer: remove the token before storage, derive it again later, insert it back into the lifecycle payload, and only then call `createSession()`.

The current examples configure deterministic minting but persist/pass the returned resume payload unchanged, so they continue carrying the token. The finding is therefore a missing public reconstruction path rather than evidence that deterministic minting itself fails.

## Current behavior

The Codex live-bridge lifecycle schema requires:

```text
bridge.port
bridge.token
bridge.lastSeenEventId
bridge.sandboxId? 
```

`doDetach()` and `doSuspendTurn()` both return live bridge coordinates containing:

```text
port
bridgeToken
lastSeenEventId
sandboxId
```

On live attach the adapter constructs the WebSocket URL from `coords.token`. The configured mint callback is used for spawn/respawn and is not called on the existing attach path.

Before any adapter logic runs, `HarnessAgent.createSession()` validates `resumeFrom` / `continueFrom` against the harness lifecycle schema. A payload with `bridge.token` removed therefore cannot reach an adapter path that might derive the token.

## Target-native discriminator

Owned source: `teamleaderleo/ai#51@c40f7b2a5de2a6ad6dcddee8f7b3864a00fc5358`  
Execution carrier: `teamleaderleo/ai#52@b51060b709b14967e8ae4d4c0f89b54bbbe276c4`  
Run: `31423683026`  
Runner: Ubuntu 24.04 / Node 22

The test uses the real Codex lifecycle schema, `HarnessAgent.createSession()`, deterministic `mintBridgeToken`, and a sentinel sandbox provider.

Observed result:

1. Remove `bridge.token` from valid live resume coordinates.
2. Call `createSession()` with deterministic mint configured.
3. Lifecycle validation rejects before `resumeSession()` is called; the mint callback is also untouched.
4. Insert the derived token into the same state.
5. Validation succeeds and execution reaches the sandbox resume sentinel.

Exact package receipt:

```text
@ai-sdk/harness-codex type-check: PASS
Fieldwork characterization: PASS
ordinary package tests beside it: 83 PASS
Fieldwork test: 1 PASS
84 / 84 total PASS
```

Evidence class: `target-executed` for the public validation/rehydration contract.

## Introducing-change context

The introducing change says bridge-backed harnesses previously generated a random token that had to be persisted with resume state and adds deterministic minting for callers that can derive the token from sandbox identity so they can reconstruct resume state without storing that secret.

Its adapter docs describe `mintBridgeToken` only as a function that returns a suitably secret bridge token. Its attach examples now configure deterministic minting but still retain the unmodified `resumeState` returned from `session.detach()` and pass that same object to `createSession()`.

Its unit tests explicitly assert that attach reuses the token already present in detached state and does not invoke `mintBridgeToken` again.

## Consequence

Applications persisting lifecycle state across processes can avoid storing this bearer token only by owning an undocumented transformation of an SDK-defined state object. Missing or stale rehydration fails before sandbox resume. Dropping bridge coordinates instead changes the recovery route from live attach toward respawn/replay/rerun behavior, which can weaken in-flight continuity.

This is an interface/security-ergonomics consequence. No exploit or credential exposure outside application-owned persistence is claimed.

### A — adapter-resolved optional token

Allow `bridge.token` to be absent in lifecycle schema. When live coordinates contain `sandboxId` and deterministic minting is configured, derive the token before attach. Require serialized token only when no mint callback can reconstruct it.

Question: generic schema validation cannot express the runtime setting dependency cleanly; adapters would need a clear failure when neither source exists.

### B — framework lifecycle hydration hook

Add a harness-owned lifecycle-state hydration/normalization hook that runs before schema validation. The adapter can fill derived credential fields from stable public identifiers without applications patching opaque state objects themselves.

Question: this widens the generic harness contract and may be excessive for one credential field.

### C — explicit public serializer/rehydrator

Keep strict lifecycle schemas and expose a supported helper that strips derived secrets for persistence and rehydrates them before `createSession()`.

Question: more API surface, but the ownership boundary becomes explicit and validation remains strict.

### D — retain current caller-owned pattern

Treat deterministic minting as a primitive only. Document that applications must redact and reinsert `bridge.token` around storage.

Question: the current examples and setting docs do not demonstrate that two-step contract, and callers must know an adapter-private lifecycle shape.

## Reversing evidence

Downgrade this finding if a supported public serializer/hydrator already exists elsewhere in the harness API, or if the intended contract explicitly requires applications to treat adapter lifecycle state as editable and provides a stable documented way to rebuild derived fields before validation.

## Disposition

`RETAIN / DESIGN BOUNDARY NEEDED`

No production patch is selected. The next useful step is to compare the smallest adapter-local optional-token design with a generic lifecycle hydration hook, using detach, suspended-turn replay, bridge-dead fallback, and random-token defaults as negative controls.

No third-party upstream mutation occurred.

**Comment by teamleaderleo** [comment]
Design comparison refreshed against public source `74556f7946cdf50aa41c01c5d5b3bd2b733acc86` and retained in Fieldwork PR #789 at `programmes/sdk-integration-lifecycle/scouts/vercel-ai-current-main-20260811/bridge-token-design-comparison.md`.

Provisional winner: adapter-side derived-secret resolution, not a generic pre-validation lifecycle hook.

Reason: `HarnessAgent` validates lifecycle state before sandbox acquisition, so a generic pre-validation hook does not have the authoritative resumed `sandboxSession.id`. Every bridge-backed adapter already receives that live sandbox identity inside `doStart()` immediately before attach. The narrower design permits only the derived secret to be absent, resolves `persisted token ?? mintBridgeToken(resumed sandbox id)`, preserves existing persisted-token precedence, and fails explicitly when random-token state has been redacted without a reconstruction function.

Target-native comparison controls are now explicit: deterministic detach resume, suspended in-flight continuation, bridge-dead replay/rerun fallback, persisted-token compatibility, random-token redaction failure, and resumed-sandbox identity precedence.

No production patch is selected yet. Next useful transition is one test-first adapter candidate proving those controls before deciding whether the tiny resolver deserves sharing across Codex/Claude/OpenCode/DeepAgents/ACP.

Upstream contact authorized: false.
